File: src/main/java/controller/ChangePassword.java
package controller;

import dao.CustomerDAO;
import dao.CustomerDAOImpl;
import dao.CustomerAuthDAO;
import dao.CustomerAuthDAOImpl;
import jakarta.servlet.*;
import jakarta.servlet.http.*;
import jakarta.servlet.annotation.*;
import model.Customer;
import model.CustomerAuth;
import org.apache.commons.lang3.RandomStringUtils;
import utils.HashGeneratorUtils;
import utils.RegexCheckerUtils;

import java.io.IOException;

@WebServlet(name = "ChangePassWord", value = "/change-password-servlet")
public class ChangePassword extends HttpServlet {
    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
        response.sendRedirect("user/change-password.jsp");
    }

    @Override
    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
        String currentPassword = request.getParameter("currentPassword");
        String newPassword = request.getParameter("newPassword");
        String newPasswordRetype = request.getParameter("newPasswordRetype");
        boolean logOutDevices = "true".equals(request.getParameter("logOutDevices"));
        HttpSession session = request.getSession();
        String url = "user/change-password.jsp"; // default url
        Customer customer = (Customer) session.getAttribute("loggedCustomer");

        // check if the user is logged in
        if (customer == null) {
            session.setAttribute("error", "You need to login to change password!");
            response.sendRedirect("login.jsp");
            return;
        }

        String currentPasswordHash = HashGeneratorUtils.generateSHA256DefaultSalt(currentPassword);

        // get the customer from the database because the password might be changed by another device
        // and the session object might not be updated
        Customer customerFromDB = new CustomerDAOImpl().selectById(customer);

        // update the session object
        session.setAttribute("loggedCustomer", customerFromDB);

        // get the password hash from the database
        String databasePasswordHash = customerFromDB.getPassword();

        boolean hasError = false;
        if (!newPassword.equals(newPasswordRetype)) {
            session.setAttribute("errorNewPasswordRetype", "Password does not match! Enter new password again here.");
            hasError = true;
        }
        if (!currentPasswordHash.equals(databasePasswordHash)) {
            session.setAttribute("errorCurrentPassword", "Current password is incorrect. Please try again.");
            hasError = true;
        }
        if (!RegexCheckerUtils.checkPassword(newPassword)) {
            session.setAttribute("errorNewPassword", "Choose a more secure password. It should be at least 8 characters and difficult for others to guess.");
            hasError = true;
        }
        if (newPassword.equals(currentPassword)) {
            session.setAttribute("errorNewPassword", "New password must be different from the current password.");
            hasError = true;
        }

        if (hasError) {
            session.setAttribute("currentPassword", currentPassword);
            session.setAttribute("newPassword", newPassword);
            session.setAttribute("newPasswordRetype", newPasswordRetype);
        } else {
            customer.setPassword(HashGeneratorUtils.generateSHA256DefaultSalt(newPassword));
            session.setAttribute("loggedCustomer", customer);
            CustomerDAO customerDAO = new CustomerDAOImpl();
            // if the password is updated successfully
            if (customerDAO.updatePassword(customer)) {
                session.setAttribute("message", "Password changed successfully!");
                // implement the log out all devices feature
                if (logOutDevices) {
//                    session.removeAttribute("loggedCustomer");
//                    session.invalidate();
                    CustomerAuthDAO customerAuthDAO = new CustomerAuthDAOImpl();
                    customerAuthDAO.deleteAllTokens(customer); // delete all tokens from the database

                    // create a secret token to send request to logout-all servlet
                    String token = RandomStringUtils.randomAlphanumeric(12);
                    session.setAttribute("token", token);
                    url = "/logout-all" + "?token=" + token;
                }
            } else {
                session.setAttribute("message", "An error occurred while changing password. Please try again.");
            }
        }

        response.sendRedirect(url);
    }
}

File: src/main/webapp/assets/js/change-password.js
const currentPassword = document.getElementById('currentPassword');
const newPassword = document.getElementById('newPassword');
const newPasswordRetype = document.getElementById('newPasswordRetype');
const errorCurrentPassword = document.getElementById('errorCurrentPassword');
const errorNewPassword = document.getElementById('errorNewPassword');
const errorNewPasswordRetype = document.getElementById('errorNewPasswordRetype');
const changePasswordBtn = document.getElementById('changePasswordBtn');
const formFloatingList = document.getElementsByClassName('form-floating');

const passwordRegex = /^(?=.*?[A-Z])(?=.*?[a-z])(?=.*?[0-9])(?=.*?[#?!@$%^&*-]).{8,}$/;

let isPasswordMatch = newPassword && newPassword.value === newPasswordRetype.value;
let isPasswordSecure = newPassword && passwordRegex.test(newPassword.value);
let isPasswordEmpty = true;

const passwordElements = [currentPassword, newPassword, newPasswordRetype];
const newPasswordElements = [newPassword, newPasswordRetype];
const errorElements = [errorCurrentPassword, errorNewPassword, errorNewPasswordRetype];


for (let i = 0; i < errorElements.length; i++) {
    if (errorElements[i].innerHTML) {
        formFloatingList[i].classList.add('is-invalid');
        passwordElements[i].classList.add('is-invalid');
    }
}

currentPassword.addEventListener('keyup', () => {
    if (currentPassword.value && currentPassword.value.length < 8) {
        errorCurrentPassword.innerHTML = "Password is too short!";
        formFloatingList[0].classList.add('is-invalid');
        currentPassword.classList.add('is-invalid');
    } else {
        errorCurrentPassword.innerHTML = "";
        formFloatingList[0].classList.remove('is-invalid');
        currentPassword.classList.remove('is-invalid');
    }
});

currentPassword.addEventListener('keyup', () => {
    errorCurrentPassword.innerHTML = "";
    formFloatingList[0].classList.remove('is-invalid');
    currentPassword.classList.remove('is-invalid');
});

newPasswordElements.forEach((element) => {
    element.addEventListener('keyup', () => {
        if (newPasswordRetype.value && newPassword.value !== newPasswordRetype.value) {
            errorNewPasswordRetype.innerHTML = "Password does not match! Enter new password again here.";
            formFloatingList[2].classList.add('is-invalid');
            newPasswordRetype.classList.add('is-invalid');
            isPasswordMatch = false;
        } else {
            errorNewPasswordRetype.innerHTML = "";
            formFloatingList[2].classList.remove('is-invalid');
            newPasswordRetype.classList.remove('is-invalid');
            isPasswordMatch = true;
        }
    });
});

newPassword.addEventListener('keyup', () => {
    if (newPassword.value && !passwordRegex.test(newPassword.value)) {
        errorNewPassword.innerHTML = "Choose a more secure password that you don't use anywhere else. It should be at least 8 characters and difficult for others to guess.";
        formFloatingList[1].classList.add('is-invalid');
        newPassword.classList.add('is-invalid');
        isPasswordSecure = false;
    } else {
        errorNewPassword.innerHTML = "";
        formFloatingList[1].classList.remove('is-invalid');
        newPassword.classList.remove('is-invalid');
        isPasswordSecure = true;
    }
});

function checkEmptyPassword() {
    let isEmpty = false;
    passwordElements.forEach((element) => {
        if (!element.value) {
            isEmpty = true;
        }
    });
    return isEmpty;
}

passwordElements.forEach((element) => {
    element.addEventListener('keyup', () => {
        isPasswordEmpty = checkEmptyPassword();
        changePasswordBtn.disabled = isPasswordEmpty || !isPasswordMatch || !isPasswordSecure;
    });
});